org.opentosca.planbuilder/src/org/opentosca/planbuilder/AbstractTransformingPlanbuilder.java

```diff
@@ -247,28 +359,71 @@ private boolean contains(Collection<AbstractNodeTemplate> subgraph1, Collection<
                                                            Set<AbstractNodeTemplate> graph2,
                                                            Set<AbstractNodeTemplate> currentSubset) {
 
+        LOG.debug("Finding MaxCommon Subgraph with vertices {}", this.printCandidate(vertices));
         if (vertices.isEmpty()) {
             if (this.isCommonSubgraph(graph1, graph2, currentSubset)) {
+                LOG.debug("Returning the current subset of {}" , this.printCandidate(currentSubset));
                 return new HashSet<AbstractNodeTemplate>(currentSubset);
             } else {
                 return new HashSet<AbstractNodeTemplate>();
             }
         }
 
+        
+        
         AbstractNodeTemplate v = this.pop(vertices);
-        currentSubset.add(v);
+        
+        LOG.debug("Removed vertex {}" , v.getId());        
         Set<AbstractNodeTemplate> cand1 = this.getMaxCommonSubgraph(vertices, graph1, graph2, currentSubset);
+        currentSubset.add(v);        
+        LOG.debug("Current subset {}" , this.printCandidate(currentSubset));
+        
+        
         Set<AbstractNodeTemplate> cand2 = new HashSet<AbstractNodeTemplate>();
 
         if (this.isCommonSubgraph(graph1, graph2, currentSubset)) {
             cand2 = this.getMaxCommonSubgraph(vertices, graph1, graph2, currentSubset);
+        } else {
+            LOG.debug("Removing vertex {} from current subset {}", v.getId(), this.printCandidate(currentSubset));            
+            currentSubset.remove(v);
         }
+        
 
-        currentSubset.remove(v);
 
+        LOG.debug("Current candidates:");
+        LOG.debug("Candidate1: {}" , this.printCandidate(cand1));
+        LOG.debug("Candidate2: {}" , this.printCandidate(cand2));
+        
+        if(cand1.size() > cand2.size()) {            
+            LOG.debug("Returning cand1");
+        } else {            
+            LOG.debug("Returning cand2");
+        }
+        
         return (cand1.size() > cand2.size()) ? cand1 : cand2;
     }
 
+    private String printCandidate(Collection<AbstractNodeTemplate> nodeTemplates) {
+        String print = "{";
+        
+        
+        AbstractNodeTemplate[] nodes = nodeTemplates.toArray(new AbstractNodeTemplate[nodeTemplates.size()]);
+        
+        
+        for(int i = 0 ; i < nodes.length; i++) {
+            
+            print += nodes[i].getId();
+            
+            if(i+1 < nodes.length) {
+                print += ",";
+            }
+        }
+
+        print += "}";
+        
+        return print;
```

Comment:
instead of this, how about we use:

```
return "{" + nodeTemplates.stream()
    .map(AbstractNodeTemplate::getId)
    .collect(Collectors.joining(",")) + "}";
```